File: javascript/json-transform/src/JsonElementStreamer.ts
import FunctionContext from "./functions/common/FunctionContext";
import { asAsyncSequence, AsyncSequence, emptyAsyncSequence } from "@wortise/sequency";

class JsonElementStreamer {
  private readonly context: FunctionContext;
  private readonly transformed: boolean;
  private value?: any[];
  private readonly _stream?: AsyncSequence<any>;

  private constructor(context: FunctionContext, value: any[] | AsyncSequence<any>, transformed: boolean) {
    this.context = context;
    this.transformed = transformed;
    if (Array.isArray(value)) {
      this.value = value;
      this._stream = undefined;
    } else {
      this.value = undefined;
      this._stream = value;
    }
  }

  public knownAsEmpty() {
    return this.value && this.value.length == 0;
  }

  public stream(skip: number = 0, limit: number = -1) {
    if (this._stream != null && !this.value) {
      const skipped = skip > 0 ? this._stream.drop(skip) : this._stream;
      return limit > -1 ? skipped.take(limit) : skipped;
    }
    if (this.value == null) {
      return emptyAsyncSequence();
    }
    let valueStream = asAsyncSequence(this.value);
    if (skip > 0) {
      valueStream = valueStream.drop(skip);
    }
    if (limit > -1) {
      valueStream = valueStream.take(limit);
    }
    if (!this.transformed) {
      valueStream = valueStream.map(el => this.context.transform(undefined, el));
    }
    return valueStream;
  }

  public static fromJsonArray(context: FunctionContext, value: any[], transformed: boolean) {
    return new JsonElementStreamer(context, value, transformed);
  }

  public static fromTransformedStream(context: FunctionContext, stream: AsyncSequence<any>) {
    return new JsonElementStreamer(context, stream, true);
  }

  public async toJsonArray() {
    if (this.value) {
      return this.value;
    }
    if (this._stream) {
      return this._stream.toArray().then(arr => {
        this.value = arr;
        return arr;
      });
    }
    return [];
  }
}

export default JsonElementStreamer;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/JsonElementStreamer.java
package co.nlighten.jsontransform;

import co.nlighten.jsontransform.functions.common.FunctionContext;

import java.util.stream.Stream;

public class JsonElementStreamer {
    private final FunctionContext context;
    private final boolean transformed;
    private Object value;
    private final Stream<?> stream;

    private JsonElementStreamer(FunctionContext context, Stream<?> stream, Object arr, boolean transformed) {
        this.context = context;
        this.value = arr;
        this.transformed = transformed;
        this.stream = stream;
    }

    public boolean knownAsEmpty() {
        return value != null && this.context.getAdapter().isEmpty(value);
    }

    public Stream<?> stream() {
        return stream(null, null);
    }

    public Stream<?> stream(Long skip, Long limit) {
        if (this.stream != null && this.value == null) {
            var skipped = skip != null ? this.stream.skip(skip) : this.stream;
            return limit != null ? skipped.limit(limit) : skipped;
        }
        if (value == null) {
            return Stream.empty();
        }
        var valueStream = context.getAdapter().stream(value, false);
        if (skip != null) {
            valueStream = valueStream.skip(skip);
        }
        if (limit != null) {
            valueStream = valueStream.limit(limit);
        }
        if (!transformed) {
            valueStream = valueStream.map(context::transform);
        }
        return valueStream;
    }

    public static JsonElementStreamer fromJsonArray(
            FunctionContext context, Object array, boolean transformed) {
        return new JsonElementStreamer(context, null, array, transformed);
    }

    public static JsonElementStreamer fromTransformedStream(
            FunctionContext context, Stream<?> stream) {
        return new JsonElementStreamer(context, stream, null, true);
    }

    public Object toJsonArray() {
        if (value != null) {
            return value;
        }
        var adapter = context.getAdapter();
        var ja = adapter.createArray();
        if (stream != null) {
            stream.forEach(item -> adapter.add(ja, item));
        }
        value = ja;
        return ja;
    }
}
